<a href="https://colab.research.google.com/github/scoobiii/ONS_Dados_Abertos/blob/main/SunStream_SIN_GERACAO_USINA_2021.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# GERACAO USINA 2021 

In [2]:
from urllib import request
import pandas as  pd
import numpy as np

**1. Download do Dataset do Portal de Dados Abertos**





In [4]:
file_url = 'https://ons-dl-prod-opendata.s3.amazonaws.com/dataset/geracao_usina_ho/GERACAO_USINA_2021.csv'
source_file = 'GERACAO_USINA_2021.csv'
request.urlretrieve(file_url , source_file )

('GERACAO_USINA_2021.csv', <http.client.HTTPMessage at 0x7f0c85ca0be0>)

**2. Importar arquivo**

In [5]:
dataframe = pd.read_csv(source_file, sep=';',encoding = "utf-8")

if (dataframe.empty) :
    print('DATAFRAME VAZIO')
else:
  print('shape:', dataframe.shape)
  print('columns:', dataframe.columns)

shape: (4868160, 11)
columns: Index(['din_instante', 'id_subsistema', 'nom_subsistema', 'id_estado',
       'nom_estado', 'cod_modalidadeoperacao', 'nom_tipousina',
       'nom_tipocombustivel', 'nom_usina', 'ceg', 'val_geracao'],
      dtype='object')


In [6]:
dataframe.head(3)

din_instante id_subsistema nom_subsistema id_estado nom_estado  \
0  2021-01-01 00:00:00             N          NORTE        AM   AMAZONAS   
1  2021-01-01 00:00:00             N          NORTE        AM   AMAZONAS   
2  2021-01-01 00:00:00             N          NORTE        AM   AMAZONAS   

  cod_modalidadeoperacao  nom_tipousina nom_tipocombustivel nom_usina  \
0                 TIPO I  HIDROELÉTRICA          Hidráulica   Balbina   
1               TIPO III        TÉRMICA         Óleo Diesel  Flores 1   
2               TIPO III        TÉRMICA         Óleo Diesel  Flores 4   

                     ceg  val_geracao  
0  UHE.PH.AM.000190-2.01      106.554  
1  UTE.PE.AM.029192-7.01        0.000  
2  UTE.PE.AM.029192-7.01        0.000

In [7]:
dataframe.head()

din_instante id_subsistema nom_subsistema id_estado nom_estado  \
0  2021-01-01 00:00:00             N          NORTE        AM   AMAZONAS   
1  2021-01-01 00:00:00             N          NORTE        AM   AMAZONAS   
2  2021-01-01 00:00:00             N          NORTE        AM   AMAZONAS   
3  2021-01-01 00:00:00             N          NORTE        AM   AMAZONAS   
4  2021-01-01 00:00:00             N          NORTE        AM   AMAZONAS   

  cod_modalidadeoperacao  nom_tipousina nom_tipocombustivel        nom_usina  \
0                 TIPO I  HIDROELÉTRICA          Hidráulica          Balbina   
1               TIPO III        TÉRMICA         Óleo Diesel         Flores 1   
2               TIPO III        TÉRMICA         Óleo Diesel         Flores 4   
3                 TIPO I        TÉRMICA                 Gás        Aparecida   
4                 TIPO I        TÉRMICA                 Gás  Cristiano Rocha   

                     ceg  val_geracao  
0  UHE.PH.AM.000190-2.01      106.554  
1  UTE.PE.AM.029192-7.01        0.000  
2  UTE.PE.AM.029192-7.01        0.000  
3  UTE.GN.AM.027250-7.02       91.630  
4  UTE.PE.AM.029499-3.01       65.842

In [8]:
dataframe

din_instante id_subsistema nom_subsistema id_estado  \
0        2021-01-01 00:00:00             N          NORTE        AM   
1        2021-01-01 00:00:00             N          NORTE        AM   
2        2021-01-01 00:00:00             N          NORTE        AM   
3        2021-01-01 00:00:00             N          NORTE        AM   
4        2021-01-01 00:00:00             N          NORTE        AM   
...                      ...           ...            ...       ...   
4868155  2021-12-31 23:00:00            SE        SUDESTE        SP   
4868156  2021-12-31 23:00:00            SE        SUDESTE        SP   
4868157  2021-12-31 23:00:00            SE        SUDESTE        TO   
4868158  2021-12-31 23:00:00            SE        SUDESTE        TO   
4868159  2021-12-31 23:00:00            SE        SUDESTE        TO   

        nom_estado      cod_modalidadeoperacao  nom_tipousina  \
0         AMAZONAS                      TIPO I  HIDROELÉTRICA   
1         AMAZONAS                    TIPO III        TÉRMICA   
2         AMAZONAS                    TIPO III        TÉRMICA   
3         AMAZONAS                      TIPO I        TÉRMICA   
4         AMAZONAS                      TIPO I        TÉRMICA   
...            ...                         ...            ...   
4868155  SAO PAULO  Pequenas Usinas (Tipo III)        TÉRMICA   
4868156  SAO PAULO                      TIPO I        TÉRMICA   
4868157  TOCANTINS                      TIPO I  HIDROELÉTRICA   
4868158  TOCANTINS                      TIPO I  HIDROELÉTRICA   
4868159  TOCANTINS                      TIPO I  HIDROELÉTRICA   

              nom_tipocombustivel                     nom_usina  \
0                      Hidráulica                       Balbina   
1                     Óleo Diesel                      Flores 1   
2                     Óleo Diesel                      Flores 4   
3                             Gás                     Aparecida   
4                             Gás               Cristiano Rocha   
...                           ...                           ...   
4868155  Outras Multi-Combustível  PQU SPSP TERM - CARGA GLOBAL   
4868156                       Gás              Nova Piratininga   
4868157                Hidráulica                  São Salvador   
4868158                Hidráulica                       Lajeado   
4868159                Hidráulica                 Peixe Angical   

                           ceg  val_geracao  
0        UHE.PH.AM.000190-2.01      106.554  
1        UTE.PE.AM.029192-7.01        0.000  
2        UTE.PE.AM.029192-7.01        0.000  
3        UTE.GN.AM.027250-7.02       91.630  
4        UTE.PE.AM.029499-3.01       65.842  
...                        ...          ...  
4868155                      -      294.000  
4868156  UTE.GN.SP.028191-3.01        0.000  
4868157  UHE.PH.TO.028567-6.01      233.109  
4868158  UHE.PH.TO.001304-8.01      584.262  
4868159  UHE.PH.TO.028353-3.01      420.073  

[4868160 rows x 11 columns]

**3.Explorando o conteúdo do arquivo**

In [9]:
dataframe.count()

din_instante              4868160
id_subsistema             4868160
nom_subsistema            4868160
id_estado                 4868160
nom_estado                4868160
cod_modalidadeoperacao    4868160
nom_tipousina             4868160
nom_tipocombustivel       4863192
nom_usina                 4868160
ceg                       4868160
val_geracao               4860888
dtype: int64

**3.2 Quantidade de valores nulos**

In [10]:
dataframe.isnull().sum()

din_instante                 0
id_subsistema                0
nom_subsistema               0
id_estado                    0
nom_estado                   0
cod_modalidadeoperacao       0
nom_tipousina                0
nom_tipocombustivel       4968
nom_usina                    0
ceg                          0
val_geracao               7272
dtype: int64

**4. Estatísticas**

In [11]:
# Variaveis do tipo "TEXTOS"
dataframe.describe(include=[object])  

din_instante id_subsistema nom_subsistema id_estado nom_estado  \
count               4868160       4868160        4868160   4868160    4868160   
unique                 8760             4              5        28         28   
top     2021-03-10 11:00:00            SE        SUDESTE        BA      BAHIA   
freq                    596       2016288        2007528    625440     625440   

       cod_modalidadeoperacao  nom_tipousina nom_tipocombustivel  nom_usina  \
count                 4868160        4868160             4863192    4868160   
unique                      7              5                  11        655   
top                    TIPO I  HIDROELÉTRICA          Hidráulica  Simplício   
freq                  2101032        1789464             1789464      17520   

            ceg  
count   4868160  
unique      370  
top           -  
freq    1980696

In [12]:
# Variaveis do tipo "NUMERO"
dataframe.describe(include=[np.number])
 
# ou dataframe.describe(exclude=[object]) 

val_geracao
count  4.860888e+06
mean   1.297298e+02
std    4.004803e+02
min    0.000000e+00
25%    1.650000e+00
50%    3.469500e+01
75%    1.078860e+02
max    1.101831e+04

In [18]:
# Estatistica sobre uma das colunas selecionadas
dataframe.val_geracao.describe()

count    4.860888e+06
mean     1.297298e+02
std      4.004803e+02
min      0.000000e+00
25%      1.650000e+00
50%      3.469500e+01
75%      1.078860e+02
max      1.101831e+04
Name: val_geracao, dtype: float64

**5. Análise sobre os resultados - Foco em Geração**

*A função value_counts() do pandas é valiosíssima nas análises de dados, o que ela faz é verificar a contagem de dados iguais em um dataframe ou em uma coluna, e retorna uma Series com esse resultado.*

In [20]:
dataframe.nom_usina.value_counts(ascending=False)

Simplício                     17520
Rondon II                     17520
São José                      16008
Itumbiara                     10584
Caçu                           8760
                              ...  
Ventos de Santa Martina 09     1152
Conj. Chafariz Norte            888
Conj. Ventos da Bahia III       696
Conj. Lagoa do Barro II         576
Guaricana                       384
Name: nom_usina, Length: 655, dtype: int64

*A função duplicated() nos retorna uma série booleana denotando linhas duplicadas. Os resultados são Verdadeiro(True) e Falso(False), sendo verdadeiros os valores duplicados.*

In [24]:
resultado = dataframe.duplicated(subset=['nom_usina'], keep=False)
dataframe[resultado]


din_instante id_subsistema nom_subsistema id_estado  \
0        2021-01-01 00:00:00             N          NORTE        AM   
1        2021-01-01 00:00:00             N          NORTE        AM   
2        2021-01-01 00:00:00             N          NORTE        AM   
3        2021-01-01 00:00:00             N          NORTE        AM   
4        2021-01-01 00:00:00             N          NORTE        AM   
...                      ...           ...            ...       ...   
4868155  2021-12-31 23:00:00            SE        SUDESTE        SP   
4868156  2021-12-31 23:00:00            SE        SUDESTE        SP   
4868157  2021-12-31 23:00:00            SE        SUDESTE        TO   
4868158  2021-12-31 23:00:00            SE        SUDESTE        TO   
4868159  2021-12-31 23:00:00            SE        SUDESTE        TO   

        nom_estado      cod_modalidadeoperacao  nom_tipousina  \
0         AMAZONAS                      TIPO I  HIDROELÉTRICA   
1         AMAZONAS                    TIPO III        TÉRMICA   
2         AMAZONAS                    TIPO III        TÉRMICA   
3         AMAZONAS                      TIPO I        TÉRMICA   
4         AMAZONAS                      TIPO I        TÉRMICA   
...            ...                         ...            ...   
4868155  SAO PAULO  Pequenas Usinas (Tipo III)        TÉRMICA   
4868156  SAO PAULO                      TIPO I        TÉRMICA   
4868157  TOCANTINS                      TIPO I  HIDROELÉTRICA   
4868158  TOCANTINS                      TIPO I  HIDROELÉTRICA   
4868159  TOCANTINS                      TIPO I  HIDROELÉTRICA   

              nom_tipocombustivel                     nom_usina  \
0                      Hidráulica                       Balbina   
1                     Óleo Diesel                      Flores 1   
2                     Óleo Diesel                      Flores 4   
3                             Gás                     Aparecida   
4                             Gás               Cristiano Rocha   
...                           ...                           ...   
4868155  Outras Multi-Combustível  PQU SPSP TERM - CARGA GLOBAL   
4868156                       Gás              Nova Piratininga   
4868157                Hidráulica                  São Salvador   
4868158                Hidráulica                       Lajeado   
4868159                Hidráulica                 Peixe Angical   

                           ceg  val_geracao  
0        UHE.PH.AM.000190-2.01      106.554  
1        UTE.PE.AM.029192-7.01        0.000  
2        UTE.PE.AM.029192-7.01        0.000  
3        UTE.GN.AM.027250-7.02       91.630  
4        UTE.PE.AM.029499-3.01       65.842  
...                        ...          ...  
4868155                      -      294.000  
4868156  UTE.GN.SP.028191-3.01        0.000  
4868157  UHE.PH.TO.028567-6.01      233.109  
4868158  UHE.PH.TO.001304-8.01      584.262  
4868159  UHE.PH.TO.028353-3.01      420.073  

[4868160 rows x 11 columns]

**6. Conclusão**

Não temos linhas repetidas no Dataset
Observamos que o resultado quantitativo da estatística para "nom_usina" teve count= 4868160 e unique= 17520	sugerindo repetição.

Numa investigação específica sobre nom_usina, encontramos 3 nomes iguais POREM para para USINAS DIFERENTES.